node/subsystem-util/src/metered.rs

```diff
@@ -0,0 +1,263 @@
+}
+
+impl<T> Clone for MeteredSender<T> {
+	fn clone(&self) -> Self {
+		Self {
+			meter: self.meter.clone(),
+			inner: self.inner.clone(),
+		}
+	}
+}
+
+impl<T> std::ops::Deref for MeteredSender<T> {
+	type Target = mpsc::Sender<T>;
+	fn deref(&self) -> &Self::Target {
+		&self.inner
+	}
+}
+
+impl<T> std::ops::DerefMut for MeteredSender<T> {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		&mut self.inner
+	}
+}
+
+impl<T> MeteredSender<T> {
+	/// Get an updated accessor object for all metrics collected.
+	pub fn meter(&self) -> &Meter {
+		&self.meter
+	}
+
+	/// Send message, wait until capacity is available.
+	pub async fn send(&mut self, item: T) -> result::Result<(), mpsc::SendError>
+	where
+		Self: Unpin,
+	{
+		self.meter.fill.fetch_add(1, Ordering::SeqCst);
+		let fut = self.inner.send(item);
+		futures::pin_mut!(fut);
+		fut.await
+	}
+
+	/// Attempt to send message or fail immediately.
+	pub fn try_send(&mut self, msg: T) -> result::Result<(), mpsc::TrySendError<T>> {
+		self.inner.try_send(msg)?;
+		self.meter.fill.fetch_add(1, Ordering::SeqCst);
+		Ok(())
+	}
+}
+
+/// A peek into the inner state of a meter.
+#[derive(Debug, Clone, Default)]
+pub struct Meter {
+	/// Name of the receiver and sender pair.
+	name: &'static str,
+	// fill state of the channel
+	fill: Arc<AtomicUsize>,
+}
+
+impl Meter {
+	/// Count the number of items queued up inside the channel.
+	pub fn queue_count(&self) -> usize {
+		// when obtaining we don't care much about off by one
+		// accuracy
+		self.fill.load(Ordering::Relaxed)
+	}
+
+	pub fn name(&self) -> &'static str {
+		self.name
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use futures::executor::block_on;
+	use futures::StreamExt;
+
+	#[derive(Clone, Copy, Debug, Default)]
+	struct Msg {
+		val: u8,
+	}
+
+	#[test]
+	fn try_send_try_next() {
+		block_on(async move {
+			let (mut tx, mut rx) = channel::<Msg>(5, "goofy");
+			let msg = Msg::default();
+			assert_eq!(rx.meter().queue_count(), 0);
+			tx.try_send(msg).unwrap();
+			assert_eq!(tx.meter().queue_count(), 1);
+			tx.try_send(msg).unwrap();
+			tx.try_send(msg).unwrap();
+			tx.try_send(msg).unwrap();
+			assert_eq!(tx.meter().queue_count(), 4);
+			rx.try_next().unwrap();
+			assert_eq!(rx.meter().queue_count(), 3);
+			rx.try_next().unwrap();
+			rx.try_next().unwrap();
+			assert_eq!(tx.meter().queue_count(), 1);
+			rx.try_next().unwrap();
+			assert_eq!(rx.meter().queue_count(), 0);
+			assert!(rx.try_next().is_err());
+		});
+	}
+
+	#[test]
+	fn with_tasks() {
+		let (mut tx, mut rx) = channel::<Msg>(5, "goofy");
+		block_on(async move {
+			futures::join!(
+				async move {
+					let msg = Msg::default();
+					assert_eq!(tx.meter().queue_count(), 0);
+					tx.try_send(msg).unwrap();
+					assert_eq!(tx.meter().queue_count(), 1);
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+				},
+				async move {
+					Delay::new(Duration::from_millis(100)).await;
```

Comment:
Delaying here will work almost all of the time, but it's not 100% reliable. If you create a mutex outside the scope of the `block_on`, move the lock into the other async block, then `lock().await` here, you have a 100% reliable synchronization mechanism, which is likely to complete faster than 100ms.

Reply:
Yeah, but good enough - it's just a test :) both futures start as ready and a couple of atomic ops will never take 100ms.

Reply:
you can probably sync on a `oneshot`?